### DISCUSSION TITLE
noob question about auto-saving onSnapshot

### DISCUSSION THREAD

**Discussion by axzr** [mobxjs/mobx-state-tree#1974] [Q&A] [answered]
Hi!

mst noob here. I've done some googling around, and I'm still stuck so I'm hoping that you might be able to help me.

atm I have a very naive trigger to save model updates. It looks like this:

```
    const afterCreate = () => {
      // this ideally shouldn't fire if the model
      // has just been loaded from the database
      self.save();

      onSnapshot(self, self.save);
    };
```
    
   This mostly works fine, but it does have the odd side effect that when the model is loaded from the database it will immediately save again, which isn't ideal. Is there a way for me to tell the 'source of creation' somehow? Or should I be testing for something different? I thought about adding a flag to the model to say what last caused it to update, but that feels hacky.
   
   Or am I just going about this the wrong way entirely?

**Comment by BrianHung** [comment, marked as answer]
Move `onSnapshot` out of `afterCreate`.

**Reply by axzr** [reply]
Hi Brian,

Thanks for your reply. As a noob I don't quite follow how this solves my problem. Could you elaborate please?

Cheers!

**Reply by antondomratchev** [reply]
Hello @axzr,

The reason you have a double save is because it is referenced in both the `afterCreate` hook as well as `onSnapshot` reaction. If you are looking to perform the `self.save()` only when the model is updated, then you only need to call `self.save()` in `onSnapshot` reaction. It is ok to have your `onSnapshot` reaction setup in the `afterCreate` this will setup the reaction for you every time this model instance is created. Take a look at the [Lifecycle documentation ](https://mobx-state-tree.js.org/overview/hooks) for a better understanding when each hook is called.  To better understand what you are getting in the `onSnapshot` reaction take a look [here](https://mobx-state-tree.js.org/concepts/listeners)

**Reply by axzr** [reply]
Hey, thanks for your reply.

I took out the initial save from the afterCreate method. I just had to move the initial save into a different method for those models created by the user. Seems to do the trick!

**Reply by coolsoftwaretyler** [reply]
Marking @BrianHung's answer as correct here. Thanks all!
